tlog-witness.md

```diff
@@ -0,0 +1,226 @@
+document are to be interpreted as described in [BCP 14][] [RFC 2119][] [RFC
+8174][] when, and only when, they appear in all capitals, as shown here.
+
+[RFC 4648]: https://www.rfc-editor.org/rfc/rfc4648.html
+[BCP 14]: https://www.rfc-editor.org/info/bcp14
+[RFC 2119]: https://www.rfc-editor.org/rfc/rfc2119.html
+[RFC 8174]: https://www.rfc-editor.org/rfc/rfc8174.html
+[RFC 6962]: https://www.rfc-editor.org/rfc/rfc6962.html
+
+## Introduction
+
+This protocol allows clients, usually logs, to obtain cosignatures from
+witnesses, making it possible to produce self-contained inclusion proofs that
+can be verified offline. When producing a new checkpoint, the log reaches out to
+witnesses to request cosignatures over it, providing a consistency proof.
+Witnesses verify that the checkpoint is consistent with the previous state of
+the tree, and return a timestamped cosignature.
+
+A witness is an entity exposing an HTTPS service identified by a name and a
+public key. Each witness is configured with a list of supported log public keys.
+For each log, uniquely identified by its origin line, the witness keeps only
+track of the latest checkpoint it observed and verified.
+
+Clients are not expected to communicate directly with the witnesses, logs and
+(sometimes) monitors are, but there is no authentication of requests beyond the
+validation of the signature on the checkpoint.
+
+Witnesses need to be operated by trusted, reliable entities with minimal churn.
+Log deployments pick a witness trust policy that gets deployed into clients and
+might be difficult or impossible to change. Developing a healthy ecosystem of
+witnesses might be the biggest hurdle to widespread tlog success. It’s
+impractical to build a separate witness ecosystem for each application, so it’s
+important that witnesses be public, interoperable, and scalable. As for
+quantity, there should be more than a few, but once reached a critical mass each
+additional witness has diminishing marginal value. Additional log deployments,
+on the other hand, don’t have diminishing returns and are a desirable goal.
+Because of this, witnesses are designed to scale well with a large number of
+rarely active logs, and to support diverse log designs, including low-latency
+and "serverless" logs.
+
+## HTTPS Interface
+
+A witness is defined by a name, a public key, and by two URL prefixes: the
+*submission prefix* for write APIs and the *monitoring prefix* for read APIs.
+
+A witness MAY use the same value for both the *submission prefix* and the
+*monitoring prefix*.
+
+If exposing the machine holding the witness key material directly to the
+Internet is undesirable, operators MAY use a [bastion][].
+
+### add-checkpoint
+
+The `add-checkpoint` call is used to submit a new checkpoint to the witness,
+along with a consistency proof from a previous checkpoint, and returns the
+cosignature.
+
+    POST <submission prefix>/add-checkpoint
```

Comment:
I recommend adding a log ID into the URL. We have done this for our witness: https://github.com/transparency-dev/witness/blob/main/api/http.go#L24

The big advantage of this is that verification of the checkpoint is a lot easier, because the witness can maintain a map of ID -> vkey. Without this, the witness has a potentially huge set of vkeys for all of the logs that it needs to attempt to open every note with. That's a potentially expensive operation that will be hard to protect against DoS apart from old-skool strategies like IP limiting.

If we could also make this ID be the same as https://github.com/transparency-dev/formats/blob/main/log/identifier.go then that would be super awesome.


Of course, the other option is to read the first line of the note without verifying it. But that's an anti-pattern :-)

Reply:
its nice to be able to filter requests based on the URL structure, in case a single server hosts multiple distinct logs.